File: s2/c12.cpp
#include <unistd.h>
#include <string>
#include "encOracle.h"
#include "modeECB.h"
#include "aes128_ecb.h"

using namespace std;

class allMightyOracle : public encOracle {
	string m_key;
	aes128_ecb m_ecb;
	static const string secret;
	public:
	allMightyOracle() : m_ecb(createRandomString(16)) {}
	virtual string encrypt(const string& input) override {
		string data(input);
		data.append(b64dec(secret));
		data = pad(data, 16);
		return m_ecb.encrypt(data);
	}
};

const string allMightyOracle::secret = "Um9sbGluJyBpbiBteSA1LjAKV2l0aCBteSByYWc"\
									   "tdG9wIGRvd24gc28gbXkgaGFpciBjYW4gYmxvdw"\
									   "pUaGUgZ2lybGllcyBvbiBzdGFuZGJ5IHdhdmluZ"\
									   "yBqdXN0IHRvIHNheSBoaQpEaWQgeW91IHN0b3A/"\
									   "IE5vLCBJIGp1c3QgZHJvdmUgYnkK";

int main() {
	srand(time(NULL));
	allMightyOracle oracle;
	string test;
	string cryptxt = oracle.encrypt(test);
	string c1;
	//	discover block size for oracle
	size_t blockLen = discoverBlockLen(oracle);
	log("[>] Cipher block len: %zu", blockLen);
	//	check if ECB
	if(!isECB(oracle, blockLen)) {
		log("[-] Non ECB cipher");
		return 1;
	}
	log("[+] ECB mode discovered");

	string decoded;
	if(!decryptECB(oracle, decoded, blockLen, cryptxt.length())) {
		log("[-] No char to fill found in alphabet!");
		log("[-] Partially decoded:");
	}
	else
		log("[+] Decoded:");
	log("%s", decoded.c_str());

	return 0;
}

File: s2/c12.py
#!/usr/bin/python3

import py_common.cryptoUtils as cu
import base64
import string
import sys

class oracle:
    def __init__(self):
        self.key = cu.createRandomString(16)
        self.secret = "Um9sbGluJyBpbiBteSA1LjAKV2l0aCBteSByYWctdG9wIGRvd24gc28gbXkgaGFpciBjYW4gYmxvdwpUaGUgZ2lybGllcyBvbiBzdGFuZGJ5IHdhdmluZyBqdXN0IHRvIHNheSBoaQpEaWQgeW91IHN0b3A/IE5vLCBJIGp1c3QgZHJvdmUgYnkK"

    def encrypt(self, plain):
        plain += base64.b64decode(self.secret)
        plain = cu.pad(plain, 16)
        return cu.encryptAES_ECB(plain, self.key)

if __name__ == "__main__":
    o = oracle()
    cryptxt = o.encrypt(b"");
    test = bytearray()
    while True:
        c1 = o.encrypt(bytes(test))
        if len(cryptxt) < len(c1):
            break
        test.append(ord('A'))
    blockLen = len(c1) - len(cryptxt)
    padLen = len(test)
    print("[>] Block len:",blockLen)
    print("[>] Padding:", padLen)
    test = b"A"*3*blockLen
    if not cu.isECB(o.encrypt(test)):
        print("[-] Non ECB oracle")
        sys.exit(1)
    print("[+] ECB mode discovered")

    fillLen = blockLen-1
    decoded = bytearray([ord("A") for _ in range(fillLen)])
    #   subtract padLen as decoding is done over printable
    while (len(decoded) - fillLen + padLen) < len(cryptxt):
        #   fillLen does not count into block count
        blockNmbr = (len(decoded)-fillLen)//blockLen
        #   prefixLen means missing characters for current block
        #   so it must decrease from blockLen-1 -> 0 for each block
        #   fillLen also have to be subtract
        #   and it happens that blockLen - 1 == fillLen
        prefixLen = (fillLen * 2 - len(decoded))%blockLen
        c1 = o.encrypt(bytes(decoded[:prefixLen]))
        c1 = c1[blockNmbr*blockLen:(blockNmbr+1)*blockLen]
        found = False
        for c in string.printable:
            test = decoded[-blockLen+1:]
            test.append(ord(c))
            c2 = o.encrypt(bytes(test))[0:blockLen]
            if c2 == c1:
                found = True
                decoded.append(ord(c))
                break
        if not found:
            print("[-] No match in alphabet")
            sys.exit(1)
    print("[>] Decoded:")
    print(decoded[fillLen:].decode('latin-1'))
